src/MongoDB/BulkWriteCommandResult.c

```
@@ -0,0 +1,482 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "bson/bson.h"
+#include "mongoc/mongoc.h"
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_long.h>
+
+#include "php_array_api.h"
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+
+#include "BSON/Document.h"
+#include "MongoDB/Server.h"
+#include "MongoDB/WriteConcernError.h"
+#include "MongoDB/WriteError.h"
+#include "BulkWriteCommandResult_arginfo.h"
+
+#define PHONGO_BULKWRITECOMMANDRESULT_CHECK_ACKNOWLEDGED(method) \
+	if (!intern->is_acknowledged) { \
+		phongo_throw_exception(PHONGO_ERROR_LOGIC, "MongoDB\\Driver\\BulkWriteCommandResult::" method "() should not be called for an unacknowledged write result"); \
+		return; \
+	}
+
+zend_class_entry* php_phongo_bulkwritecommandresult_ce;
+
+static bool php_phongo_bulkwritecommandresult_get_writeconcernerrors(php_phongo_bulkwritecommandresult_t* intern, zval* return_value)
+{
+	bson_iter_t iter;
+
+	array_init(return_value);
+
+	if (!intern->write_concern_errors) {
+		return true;
+	}
+
+	for (bson_iter_init(&iter, intern->write_concern_errors); bson_iter_next(&iter);) {
```

Not sure if it matters here, but I'll note that we don't throw an exception for an invalid iterator here. That may be irrelevant as the bson may always be valid, but I wanted to make sure this was a conscious choice.

---

This was intentional as I don't expect invalid BSON from libmongoc, but it's technically problematic since the `for` loop could proceed with calling `bson_iter_next()` after initialization failed.

Replacing this with a conditional on `bson_iter_init()` will be consistent with WriteResult.c. I'd prefer to just NOP on a false return value instead of dealing with exceptions.

On the flip side, there are some subtle changes I can make to WriteResult.c as well. If `phongo_writeerror_init()` or `phongo_writeconcernerror_init` ever fail, we should return `false` immediately instead of continuing with iteration (and risk throwing multiple exceptions).

---

Replacing this with a NOP in case of failing to initialise the iterator sounds fine to me, as well as improving handling in WriteResult.c.